### ISSUE TITLE
Rotary fails silently on rule repair_ends (addressed by creation of repair.py)

### ISSUE THREAD

**Issue by LeeBergstrand** [jmtsuji/rotary#28]
This is an issue on the ```main``` branch.

### Problem Description

It looks like the ```repair_ends``` rule fails silently at the contig circularization step, leaving an empty `repaired.fasta` file, causing the pipeline to fail later down the line at the polishing step.

```tail``` of ```rotary.log```:
```
Error in rule polish_medaka:
    jobid: 16
    input: qc_long/nanopore_qc.fastq.gz, polish/medaka_input/input.fasta
    output: polish/medaka, polish/medaka/consensus.fasta
    log: logs/polish/medaka.log (check log file(s) for error details)
    conda-env: /home/lbergstrand/rnd/hybrid_assembly_pipeline/scratch_space/conda_envs/1c71b107bd42c9288eae5ab155ebbf40_
    shell:

        medaka_consensus -i qc_long/nanopore_qc.fastq.gz -d polish/medaka_input/input.fasta -o polish/medaka           -m r941_min_sup_g507 -t 4 -b 100 > logs/polish/medaka.log 2>&1

        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Removing output files of failed job polish_medaka since they might be corrupted:
polish/medaka
Shutting down this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2023-09-29T010438.495240.snakemake.log
```

```tail``` of ```medaka.log```:
```
bcftools   1.17       1.11       True     
bgzip      1.17       1.11       True     
minimap2   2.26       2.11       True     
samtools   1.17       1.11       True     
tabix      1.17       1.11       True     
Warning: Output polish/medaka already exists, may use old results.
Aligning basecalls to draft
Creating fai index file /home/lbergstrand/rnd/hybrid_assembly_pipeline/scratch_space/assembly/end_repair/repaired.fasta.fai
[faidx] Could not build fai index /home/lbergstrand/rnd/hybrid_assembly_pipeline/scratch_space/assembly/end_repair/repaired.fasta.fai
Failed to run alignment of reads to draft.
```
```tail``` of ```assembly/end_repair/verbose.log```:
```
Making ACT files from assembly/end_repair/contigs/contig_1/contig_1.fasta /home/lbergstrand/rnd/hybrid_assembly_pipeline/scratch_space/assembly/end_repair/contigs/contig_1/L50000/assembly/assembly.fasta /home/lbergstrand/rnd/hybrid_assembly_pipeline/scratch_space/assem$
syscall: samtools faidx assembly/end_repair/contigs/contig_1/L50000/assembly/assembly.fasta
[ Fri Sep 29 05:17:40 UTC 2023 ]: End repair: 'contig_1': FAILED to linked contig ends
[ Fri Sep 29 05:17:40 UTC 2023 ]: End repair: 1 contigs could not be circularized. A partial output file including successfully circularized contigs (and no linear
```

### Problem Solution

@jmtsuji Is this fixed in the ```end_repair``` branch? Specifically, this commit --> https://github.com/jmtsuji/rotary/commit/d9471ed2f5f7869f0cefb4f68fe6237cd15a35d8. Or did I run into a unique issue? The input assembly file before was about 5 MB and one contig, so I assume the assembly succeeded. 

Also, we should ensure that if this fails, it fails at the ```end_repair``` rule rather than in the polishing rules.

**Comment by jmtsuji** [comment]
@LeeBergstrand Looks like the error is in a small typo in [Line 350 of `rotary/scripts/flye_end_repair.sh` on master](https://github.com/jmtsuji/rotary/blob/d67c1c46b759e9c7756e8bb416f5c5534a5e56db/scripts/flye_end_repair.sh#L350C5-L350C11). Your `verbose.log` file ends mid-sentence at that line. I'm not sure why the Bash script did not exit with status 1 when it hit that error --  the script header include a robust shebang line (`set -euo pipefail`). Should be possible to debug this if needed.

Anyway, this issue should be fixed in the Python ported version of the code. Once that script is ready, I'll make a PR for it into `develop` (to replace the current Bash code). A simple PR from `end_repair` won't work because it completely replaces the current Bash code with Python code, unfortunately...

For now, if you want to just continue the pipeline and bypass the end repair step, you could make a fake rule in your Snakefile that symlinks the Flye assembly to the `repaired.fasta` file. Then comment out the `repair_ends` rule. That should work for a temporary bypass.

**Comment by LeeBergstrand** [comment]
For now, I just changed the symlinking in rule ```finalize_assembly``` so that it points to the ```flye``` assembly.

```
rule finalize_assembly:
    input:
        assembly="assembly/flye/assembly.fasta",
        info="assembly/flye/assembly_info.txt"
    output:
        assembly="assembly/assembly.fasta",
        info="assembly/assembly_info.txt"
    run:   
        source_relpath = os.path.relpath(str(input.assembly),os.path.dirname(str(output.assembly)))
        os.symlink(source_relpath,str(output.assembly))

        source_relpath = os.path.relpath(str(input.info),os.path.dirname(str(output.info)))
        os.symlink(source_relpath,str(output.info))
```

This causes Snakemake, to skip the ```assembly_end_repair``` rule when making the ```DAG``` because the files produced by the rule are no longer needed.

**Comment by jmtsuji** [comment]
Sounds good -- this workaround also works

**Comment by jmtsuji** [comment]
I think this should be fixed now in the end_repair branch, now that the code has been migrated to Python. I will close this issue.
